File: app/src/main/java/com/adonai/manman/ManPageDialogFragment.java
package com.adonai.manman;

import android.annotation.SuppressLint;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.provider.Browser;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v4.content.LocalBroadcastManager;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ViewFlipper;

import com.adonai.manman.database.DbProvider;
import com.adonai.manman.entities.ManPage;
import com.adonai.manman.misc.AbstractNetworkAsyncLoader;
import com.adonai.manman.parser.Man2Html;
import com.adonai.manman.views.PassiveSlidingPane;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;

import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
import org.apache.commons.compress.utils.IOUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URISyntaxException;
import java.util.Set;
import java.util.TreeSet;
import java.util.zip.GZIPInputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

/**
 * Dialog fragment for showing web page with man content
 * Retrieves info from DB (if cached) or network (if not)
 *
 * @see com.adonai.manman.entities.ManPage
 * @author Oleg Chernovskiy
 */
public class ManPageDialogFragment extends Fragment {
    private static final String USER_LEARNED_SLIDER = "user.learned.slider";

    private static final String PARAM_ADDRESS = "param.address";
    private static final String PARAM_NAME = "param.name";

    private RetrieveManPageCallback manPageCallback = new RetrieveManPageCallback();
    private File mLocalArchive;
    private String mAddressUrl;
    private String mCommandName;

    private LinearLayout mLinkContainer;
    private PassiveSlidingPane mSlider;
    private ViewFlipper mFlipper;
    private WebView mContent;

    private LinearLayout mSearchContainer;
    private EditText mSearchEdit;
    private ImageView mCloseSearchBar;
    private ImageView mFindNext;
    private ImageView mFindPrevious;

    @NonNull
    public static ManPageDialogFragment newInstance(@NonNull String commandName, @NonNull String address) {
        ManPageDialogFragment fragment = new ManPageDialogFragment();
        Bundle args = new Bundle();
        args.putString(PARAM_ADDRESS, address);
        args.putString(PARAM_NAME, commandName);
        fragment.setArguments(args);
        return fragment;
    }

    public ManPageDialogFragment() {
        // mandatory empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if(getArguments() != null) {
            mAddressUrl = getArguments().getString(PARAM_ADDRESS);
            mCommandName = getArguments().getString(PARAM_NAME);
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        setHasOptionsMenu(true);

        mLocalArchive = new File(getActivity().getCacheDir(), "manpages.zip");

        View root = inflater.inflate(R.layout.fragment_man_page_show, container, false);
        mLinkContainer = (LinearLayout) root.findViewById(R.id.link_list);
        mFlipper = (ViewFlipper) root.findViewById(R.id.flipper);
        mContent = (WebView) root.findViewById(R.id.man_content_web);
        mContent.setWebViewClient(new ManPageChromeClient());
        mContent.getSettings().setJavaScriptEnabled(true);
        mSlider = (PassiveSlidingPane) root.findViewById(R.id.sliding_pane);
        mSlider.setTrackedView(mContent);

        // search-specific
        mSearchContainer = (LinearLayout) root.findViewById(R.id.search_bar_layout);
        mSearchEdit = (EditText) mSearchContainer.findViewById(R.id.search_edit);
        mCloseSearchBar = (ImageView) mSearchContainer.findViewById(R.id.close_search_bar);
        mFindNext = (ImageView) mSearchContainer.findViewById(R.id.find_next_occurrence);
        mFindPrevious = (ImageView) mSearchContainer.findViewById(R.id.find_previous_occurrence);

        mCloseSearchBar.setOnClickListener(new SearchBarCloser());
        mSearchEdit.addTextChangedListener(new SearchExecutor());
        mFindNext.setOnClickListener(new SearchFurtherExecutor(true));
        mFindPrevious.setOnClickListener(new SearchFurtherExecutor(false));

        // Lollipop blocks mixed content but we should load CSS from filesystem
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            mContent.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
        }
        getLoaderManager().initLoader(MainPagerActivity.MAN_PAGE_RETRIEVER_LOADER, null, manPageCallback);
        return root;
    }

    @Override
    public void onResume() {
        super.onResume();
        // hide keyboard on fragment show, window token is hopefully present at this moment
        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(mContent.getWindowToken(), 0);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.man_page_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.show_search_bar:
                mSearchContainer.setVisibility(View.VISIBLE);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Class for creating a loader that performs async loading of man page from www.mankier.com
     * On finish passes data to web content and makes it active
     * On fail dismisses parent dialog
     *
     */
    private class RetrieveManPageCallback implements LoaderManager.LoaderCallbacks<ManPage> {
        @NonNull
        @Override
        public Loader<ManPage> onCreateLoader(int id, Bundle args) {
            return new AbstractNetworkAsyncLoader<ManPage>(getActivity()) {

                @Nullable
                @Override
                public ManPage loadInBackground() {
                    // handle special case when it's a local file
                    if(mAddressUrl.startsWith("/")) { // TODO: rewrite with URI
                        try {
                            File input = new File(mAddressUrl);
                            String charset = Utils.detectEncodingOfArchive(input);
                            FileInputStream fis = new FileInputStream(input);
                            GZIPInputStream gis = new GZIPInputStream(fis);
                            BufferedReader br = charset == null ? new BufferedReader(new InputStreamReader(gis)) : new BufferedReader(new InputStreamReader(gis, charset));
                            Man2Html parser = new Man2Html(br);
                            Document parsed = parser.getDoc();
                            ManPage result = new ManPage(input.getName(), "file://" + mAddressUrl);
                            result.setLinks(getLinks(parsed.select("div.man-page").first()));
                            result.setWebContent(parsed.html());
                            br.close(); // closes all the IS hierarchy
                            return result;
                        } catch (FileNotFoundException e) {
                            Log.e(Utils.MM_TAG, "File with man page was not found in local folder", e);
                            Toast.makeText(getActivity(), R.string.file_not_found, Toast.LENGTH_SHORT).show();
                        } catch (IOException e) {
                            Log.e(Utils.MM_TAG, "Exception while loading man page file from local folder", e);
                            Toast.makeText(getActivity(), R.string.wrong_file_format, Toast.LENGTH_SHORT).show();
                        }
                        return null; // no further querying
                    }

                    if(mAddressUrl.startsWith("local:")) { // local man archive
                        try {
                            ZipFile zip = new ZipFile(mLocalArchive);
                            ZipEntry zEntry = zip.getEntry(mAddressUrl.substring(7));
                            InputStream is =  zip.getInputStream(zEntry);
                            // can't use java's standard GZIPInputStream around zip IS because of inflating issue
                            GzipCompressorInputStream gis = new GzipCompressorInputStream(is); // manpage files are .gz
                            BufferedReader br = new BufferedReader(new InputStreamReader(gis));
                            Man2Html parser = new Man2Html(br);
                            Document parsed = parser.getDoc();
                            ManPage result = new ManPage(zEntry.getName(), mAddressUrl);
                            result.setLinks(getLinks(parsed.select("div.man-page").first()));
                            result.setWebContent(parsed.html());
                            br.close(); // closes all the IS hierarchy
                            return result;
                        } catch (IOException e) {
                            Log.e(Utils.MM_TAG, "Error while loading man page from local archive", e);
                            Toast.makeText(getActivity(), R.string.wrong_file_format, Toast.LENGTH_SHORT).show();
                        }
                    }

                    try { // query cache database for corresponding command
                        ManPage cached = DbProvider.getHelper().getManPagesDao().queryForId(mAddressUrl);
                        if(cached != null) {
                            return cached;
                        }
                    } catch (RuntimeException e) { // it's RuntimeExceptionDao, so catch runtime exceptions
                        Log.e(Utils.MM_TAG, "Exception while querying the DB", e);
                        Utils.showToastFromAnyThread(getActivity(), R.string.database_retrieve_error);
                    }

                    try {
                        OkHttpClient client = new OkHttpClient();
                        Request request = new Request.Builder().url(mAddressUrl).build();
                        Response response = client.newCall(request).execute();
                        if (response.isSuccessful()) {
                            String result = response.body().string();
                            Document root = Jsoup.parse(result, mAddressUrl);
                            Element man = root.select("div.man-page").first();
                            if (man == null) { // not actually a man page
                                return null;
                            }

                            String webContent = man.html();
                            TreeSet<String> linkContainer = getLinks(man);

                            // save to DB for caching
                            ManPage toCache = new ManPage(mCommandName, mAddressUrl);
                            toCache.setLinks(linkContainer);
                            toCache.setWebContent(webContent);
                            DbProvider.getHelper().getManPagesDao().createIfNotExists(toCache);
                            LocalBroadcastManager.getInstance(getActivity()).sendBroadcast(new Intent(MainPagerActivity.DB_CHANGE_NOTIFY));

                            return toCache;
                        }
                    } catch (IOException e) {
                        Log.e(Utils.MM_TAG, "Exception while saving cached page to DB", e);
                        // can't show a toast from a thread without looper
                        Utils.showToastFromAnyThread(getActivity(), R.string.connection_error);
                    }
                    return null;
                }

                // retrieve link set from manpage
                @NonNull
                private TreeSet<String> getLinks(Element man) {
                    Elements links = man.select("a[href*=#]");
                    TreeSet<String> linkContainer = new TreeSet<>();
                    for (Element link : links) {
                        if (!TextUtils.isEmpty(link.text()) && link.attr("href").contains("#" + link.text())) { // it's like <a href="http://ex.com/#a">-x</a>
                            linkContainer.add(link.text());
                        }
                    }
                    return linkContainer;
                }
            };
        }

        @Override
        public void onLoadFinished(Loader<ManPage> loader, ManPage data) {
            if(data != null) {
                mContent.loadDataWithBaseURL(mAddressUrl, Utils.getWebWithCss(getActivity(), data.getUrl(), data.getWebContent()), "text/html", "UTF-8", null);
                mContent.setBackgroundColor(Utils.getThemedValue(getActivity(), R.attr.fill_color)); // prevent flickering
                fillLinkPane(data.getLinks());
                mFlipper.showNext();
                shakeSlider();
            } else {
                mContent.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        getFragmentManager().popBackStack(); // can't perform transactions from onLoadFinished
                    }
                }, 0);
            }
        }

        @Override
        public void onLoaderReset(Loader<ManPage> loader) {
            // never used
        }
    }

    private void shakeSlider() {
        if(mLinkContainer.getChildCount() == 0) // nothing to show in the links pane
            return;

        SharedPreferences mPrefs = PreferenceManager.getDefaultSharedPreferences(getActivity());
        if(mPrefs.contains(USER_LEARNED_SLIDER))
            return;

        mSlider.postDelayed(new Runnable() {
            @Override
            public void run() {
                mSlider.openPane();
            }
        }, 1000);
        mSlider.postDelayed(new Runnable() {
            @Override
            public void run() {
                mSlider.closePane();
            }
        }, 2000);
        mPrefs.edit().putBoolean(USER_LEARNED_SLIDER, true).apply();
    }

    private void fillLinkPane(Set<String> links) {
        mLinkContainer.removeAllViews();

        if(links == null || links.isEmpty())
            return;

        for (final String link : links) {
            // hack  for https://code.google.com/p/android/issues/detail?id=36660 - place inside of FrameLayout
            View root = LayoutInflater.from(getActivity()).inflate(R.layout.link_text_item, mLinkContainer, false);
            TextView linkLabel = (TextView) root.findViewById(R.id.link_text);
            linkLabel.setText(link);
            root.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    mContent.loadUrl("javascript:(function(){" +
                            "l=document.querySelector('a[href$=\"#" + link + "\"]');" +
                            "e=document.createEvent('HTMLEvents');" +
                            "e.initEvent('click',true,true);" +
                            "l.dispatchEvent(e);" +
                            "})()");
                }
            });
            mLinkContainer.addView(root);
        }
    }

    /**
     * Class to load URLs inside of already active webview
     * Calls original browser intent for the URLs it can't handle
     */
    private class ManPageChromeClient extends WebViewClient {

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            if (url.matches("https://www\\.mankier\\.com/.+/.+")) { // it's an address of the command
                mFlipper.showPrevious();
                mAddressUrl = url;
                mCommandName = url.substring(url.lastIndexOf('/') + 1);
                getLoaderManager().getLoader(MainPagerActivity.MAN_PAGE_RETRIEVER_LOADER).onContentChanged();
                return true;
            }
            return shouldOverrideUrlLoadingOld(view, url);
        }

        /**
         * Copied from WebViewContentsClientAdapter (internal android class)
         * to handle URLs in old way if it's not a man page
         */
        public boolean shouldOverrideUrlLoadingOld(WebView view, String url) {
            Intent intent;
            // Perform generic parsing of the URI to turn it into an Intent.
            try {
                intent = Intent.parseUri(url, Intent.URI_INTENT_SCHEME);
            } catch (URISyntaxException ex) {
                Log.w("WebViewCallback", "Bad URI " + url + ": " + ex.getMessage());
                return false;
            }
            // Sanitize the Intent, ensuring web pages can not bypass browser
            // security (only access to BROWSABLE activities).
            intent.addCategory(Intent.CATEGORY_BROWSABLE);
            intent.setComponent(null);

            // Pass the package name as application ID so that the intent from the
            // same application can be opened in the same tab.
            intent.putExtra(Browser.EXTRA_APPLICATION_ID, view.getContext().getPackageName());
            try {
                view.getContext().startActivity(intent);
            } catch (ActivityNotFoundException ex) {
                Log.w("WebViewCallback", "No application can handle " + url);
                return false;
            }
            return true;
        }
    }

    /**
     * Closes the search bar
     */
    private class SearchBarCloser implements View.OnClickListener {
        @Override
        public void onClick(View v) {
            mSearchContainer.setVisibility(View.GONE);
            mContent.clearMatches();
        }
    }

    /**
     * Finds next occurrence depending on direction
     */
    private class SearchFurtherExecutor implements View.OnClickListener {

        private boolean goDown;

        public SearchFurtherExecutor(boolean goDown) {
            this.goDown = goDown;
        }

        @Override
        public void onClick(View v) {
            mContent.findNext(goDown);
        }
    }

    /**
     * Executes search on string change
     */
    private class SearchExecutor implements TextWatcher {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @SuppressWarnings("deprecation")
        @Override
        public void afterTextChanged(Editable s) {
            // Lollipop blocks mixed content but we should load CSS from filesystem
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                mContent.findAllAsync(s.toString()); // API 16
            } else {
                mContent.findAll(s.toString());
            }
        }
    }
}

File: app/src/main/java/com/adonai/manman/ManPageDialogFragment.kt
package com.adonai.manman

import android.annotation.SuppressLint
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.provider.Browser
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Log
import android.view.*
import android.view.inputmethod.InputMethodManager
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.*
import androidx.annotation.UiThread
import androidx.annotation.WorkerThread
import androidx.appcompat.graphics.drawable.DrawerArrowDrawable
import androidx.core.graphics.ColorUtils
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.preference.PreferenceManager
import com.adonai.manman.database.DbProvider
import com.adonai.manman.databinding.FragmentManPageShowBinding
import com.adonai.manman.entities.ManPage
import com.adonai.manman.misc.resolveAttr
import com.adonai.manman.misc.showFullscreenFragment
import com.adonai.manman.parser.Man2Html
import com.adonai.manman.service.Config
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.io.*
import java.net.URISyntaxException
import java.util.*
import java.util.zip.GZIPInputStream
import java.util.zip.ZipFile

/**
 * Dialog fragment for showing web page with man content
 * Retrieves info from DB (if cached) or network (if not)
 *
 * @see ManPage
 *
 * @author Kanedias
 */
class ManPageDialogFragment : Fragment() {
    private val mPrefChangeListener = FontChangeListener()

    private lateinit var mPrefs: SharedPreferences
    private lateinit var mLocalArchive: File

    private lateinit var binding: FragmentManPageShowBinding

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setHasOptionsMenu(true)
        mLocalArchive = File(requireActivity().cacheDir, "manpages.zip")

        mPrefs = PreferenceManager.getDefaultSharedPreferences(activity)
        mPrefs.registerOnSharedPreferenceChangeListener(mPrefChangeListener)

        binding = FragmentManPageShowBinding.inflate(inflater, container, false)

        binding.manContentWeb.webViewClient = ManPageChromeClient()
        binding.manContentWeb.settings.javaScriptEnabled = true
        binding.manContentWeb.settings.minimumFontSize = Config.fontSize

        binding.slidingPane.sliderFadeColor = ColorUtils.setAlphaComponent(Utils.getThemedValue(requireContext(), R.attr.background_color), 200)
        binding.slidingPane.setTrackedView(binding.manContentWeb)

        // search-specific
        binding.closeSearchBar.setOnClickListener(SearchBarCloser())
        binding.searchEdit.addTextChangedListener(SearchExecutor())
        binding.findNextOccurrence.setOnClickListener(SearchFurtherExecutor(true))
        binding.findPreviousOccurrence.setOnClickListener(SearchFurtherExecutor(false))

        // Lollipop blocks mixed content but we should load CSS from filesystem
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            binding.manContentWeb.settings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }

        triggerLoadPageContent()

        return binding.root
    }

    @UiThread
    private fun triggerLoadPageContent() {
        val addressUrl = requireArguments().getString(PARAM_ADDRESS)!!
        val commandName = requireArguments().getString(PARAM_NAME)!!

        binding.manpageToolbar.title = commandName
        binding.manpageToolbar.subtitle = addressUrl
        binding.manpageToolbar.navigationIcon = DrawerArrowDrawable(binding.manpageToolbar.context).apply { progress = 1.0f  }
        binding.manpageToolbar.setNavigationOnClickListener { parentFragmentManager.popBackStack() }

        binding.manpageToolbar.inflateMenu(R.menu.man_page_menu)
        binding.manpageToolbar.setOnMenuItemClickListener { mi ->
            return@setOnMenuItemClickListener when (mi.itemId) {
                R.id.show_search_bar -> {
                    toggleSearchBar(View.VISIBLE)
                    true
                }
                else -> false
            }
        }

        lifecycleScope.launch {
            val manpage = withContext(Dispatchers.IO) { doLoadContent(addressUrl, commandName) }
            if (manpage != null) {
                binding.manContentWeb.loadDataWithBaseURL(addressUrl, Utils.getWebWithCss(requireContext(), manpage.url, manpage.webContent), "text/html", "UTF-8", null)
                binding.manContentWeb.setBackgroundColor(Utils.getThemedValue(requireContext(), R.attr.background_color)) // prevent flickering
                fillLinkPane(manpage.links)

                // show the actual content on web page
                binding.manpageWaitFlipper.showNext()
                shakeSlider()
            } else {
                parentFragmentManager.popBackStack()
            }
        }
    }

    @WorkerThread
    private fun doLoadContent(addressUrl: String, commandName: String): ManPage? {
        // handle special case when it's a filesystem
        if (addressUrl.startsWith("/")) { // TODO: rewrite with URI
            try {
                val input = File(addressUrl)
                val charset = Utils.detectEncodingOfArchive(input)
                val fis = FileInputStream(input)
                val gis = GZIPInputStream(fis)

                val reader = when (charset) {
                    null -> BufferedReader(InputStreamReader(gis))
                    else -> BufferedReader(InputStreamReader(gis, charset))
                }

                reader.use {
                    val parser = Man2Html(it)
                    val parsed = parser.doc
                    val result = ManPage(input.name, "file://$addressUrl")
                    result.links = getLinks(parsed.select("div.man-page").first())
                    result.webContent = parsed.html()
                    return result
                }
            } catch (e: FileNotFoundException) {
                Log.e(Utils.MM_TAG, "File with man page was not found in local folder: $addressUrl", e)
                Toast.makeText(activity, R.string.file_not_found, Toast.LENGTH_SHORT).show()
            } catch (e: IOException) {
                Log.e(Utils.MM_TAG, "Exception while loading man page file from local folder: $addressUrl", e)
                Toast.makeText(activity, R.string.wrong_file_format, Toast.LENGTH_SHORT).show()
            }

            // no further querying, request was for local file
            return null
        }

        // special case: local file
        if (addressUrl.startsWith("local:")) {
            // local man archive
            try {
                val zip = ZipFile(mLocalArchive)
                var zEntry = zip.getEntry(addressUrl.substringAfter("local:/"))
                if (zEntry == null) {
                    // local links from the page itself can be without gz prefix
                    zEntry = zip.getEntry(addressUrl.substringAfter("local:/") + ".gz")
                }
                if (zEntry == null) {
                    // still null, reference to a non-existing page?
                    throw FileNotFoundException("File $addressUrl not found in local zip archive")
                }

                val zStream = zip.getInputStream(zEntry)
                // can't use java's standard GZIPInputStream around zip IS because of inflating issue
                val gis = GzipCompressorInputStream(zStream) // manpage files are .gz
                val reader = BufferedReader(InputStreamReader(gis))

                reader.use {
                    val parser = Man2Html(it)
                    val parsed = parser.doc
                    val result = ManPage(zEntry.name, addressUrl)
                    result.links = getLinks(parsed.select("div.man-page").first())
                    result.webContent = parsed.html()
                    return result
                }
            } catch (e: FileNotFoundException) {
                Log.e(Utils.MM_TAG, "Error while loading man page from local archive", e)
                Utils.showToastFromAnyThread(activity, getString(R.string.file_not_found_extra, addressUrl))
                return null
            } catch (e: IOException) {
                Log.e(Utils.MM_TAG, "Error while loading man page from local archive: $addressUrl", e)
                Utils.showToastFromAnyThread(activity, R.string.wrong_file_format)
                return null
            }
        }

        // query cache database for corresponding command
        val cached = DbProvider.helper.manPagesDao.queryForId(addressUrl)
        if (cached != null) {
            return cached
        }

        // exhausted all local variants, fetch remote content
        try {
            val client = OkHttpClient()
            val request = Request.Builder().url(addressUrl).build()
            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                val result = response.body!!.string()
                val root = Jsoup.parse(result, addressUrl)
                val man = root.select("div.man-page, main").first() ?: return null
                val webContent = man.html()
                val linkContainer = getLinks(man)

                // save to DB for caching
                val toCache = ManPage(commandName, addressUrl)
                toCache.links = linkContainer
                toCache.webContent = webContent
                DbProvider.helper.manPagesDao.createIfNotExists(toCache)
                LocalBroadcastManager.getInstance(requireContext()).sendBroadcast(Intent(MainPagerActivity.DB_CHANGE_NOTIFY))
                return toCache
            }
        } catch (e: IOException) {
            Log.e(Utils.MM_TAG, "Exception while saving cached page to DB: $addressUrl", e)
            Utils.showToastFromAnyThread(activity, R.string.connection_error)
        }

        return null
    }

    // retrieve link set from manpage
    private fun getLinks(man: Element): TreeSet<String> {
        val links = man.select("a[href*=#]")
        val linkContainer = TreeSet<String>()
        for (link in links) {
            if (!TextUtils.isEmpty(link.text()) && link.attr("href").contains("#" + link.text())) { // it's like <a href="http://ex.com/#a">-x</a>
                linkContainer.add(link.text())
            }
        }
        return linkContainer
    }

    override fun onResume() {
        super.onResume()
        // hide keyboard on fragment show, window token is hopefully present at this moment
        val imm = requireContext().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(binding.manContentWeb.windowToken, 0)
    }

    private fun toggleSearchBar(visibility: Int) {
        binding.searchBarLayout.visibility = visibility
        val imm = requireContext().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        if (visibility == View.VISIBLE) {
            binding.searchEdit.requestFocus()
            imm.showSoftInput(binding.searchEdit, 0)
        } else {
            binding.searchEdit.clearFocus()
            imm.hideSoftInputFromWindow(binding.searchEdit.windowToken, 0)
        }
    }

    private fun shakeSlider() {
        if (binding.linkList.childCount == 0) {
            // nothing to show in the links pane, skip
            return
        }

        if (Config.userLearnedSlider) {
            // user already seen this animation, skip
            return
        }

        binding.slidingPane.postDelayed({ binding.slidingPane.openPane() }, 1000)
        binding.slidingPane.postDelayed({ binding.slidingPane.closePane() }, 2000)

        Config.userLearnedSlider = true
    }

    private fun fillLinkPane(links: Set<String>) {
        binding.linkList.removeAllViews()
        if (links.isNullOrEmpty()) {
            return
        }

        val sortedLinks = links.sortedWith(
            compareBy<String>{ it.firstOrNull()?.isLetter()?.not() }
            .thenBy { it.length != 2 }
        )
        for (link in sortedLinks) {
            // hack  for https://code.google.com/p/android/issues/detail?id=36660 - place inside of FrameLayout
            val root = LayoutInflater.from(activity).inflate(R.layout.link_text_item, binding.linkList, false)
            val linkLabel = root.findViewById<View>(R.id.link_text) as TextView
            linkLabel.text = link
            root.setOnClickListener {
                binding.manContentWeb.loadUrl("""
                javascript:(function() {
                    var l = document.querySelector('a[href$="#$link"]');
                    var event = new MouseEvent('click', {
                        'view': window,
                        'bubbles': true,
                        'cancelable': true
                    });
                    l.dispatchEvent(event);
                })()""".trimIndent())
            }
            binding.linkList.addView(root)
        }
    }

    /**
     * Class to load URLs inside of already active webview
     * Calls original browser intent for the URLs it can't handle
     */
    private inner class ManPageChromeClient : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            if (url.matches(Regex("https://www\\.mankier\\.com/.+/.+"))) { // it's an address of the command
                val name = url.substring(url.lastIndexOf('/') + 1)

                val mpdf = newInstance(name, url)
                requireActivity().showFullscreenFragment(mpdf)

                return true
            }

            if (url.startsWith("local:")) {
                val name = url.substringAfterLast('/')

                val mpdf = newInstance(name, url)
                requireActivity().showFullscreenFragment(mpdf)

                return true
            }

            return shouldOverrideUrlLoadingOld(view, url)
        }

        /**
         * Copied from WebViewContentsClientAdapter (internal android class)
         * to handle URLs in old way if it's not a man page
         */
        fun shouldOverrideUrlLoadingOld(view: WebView, url: String): Boolean {
            // Perform generic parsing of the URI to turn it into an Intent.
            val intent = try {
                Intent.parseUri(url, Intent.URI_INTENT_SCHEME)
            } catch (ex: URISyntaxException) {
                Log.w("WebViewCallback", "Bad URI " + url + ": " + ex.message)
                return false
            }
            // Sanitize the Intent, ensuring web pages can not bypass browser
            // security (only access to BROWSABLE activities).
            intent.addCategory(Intent.CATEGORY_BROWSABLE)
            intent.component = null

            // Pass the package name as application ID so that the intent from the
            // same application can be opened in the same tab.
            intent.putExtra(Browser.EXTRA_APPLICATION_ID, view.context.packageName)
            try {
                view.context.startActivity(intent)
            } catch (ex: ActivityNotFoundException) {
                Log.w("WebViewCallback", "No application can handle $url")
                return false
            }
            return true
        }
    }

    /**
     * Closes the search bar
     */
    private inner class SearchBarCloser : View.OnClickListener {
        override fun onClick(v: View) {
            toggleSearchBar(View.GONE)
            binding.manContentWeb.clearMatches()
        }
    }

    /**
     * Finds next occurrence depending on direction
     */
    private inner class SearchFurtherExecutor(private val goDown: Boolean) : View.OnClickListener {
        override fun onClick(v: View) {
            binding.manContentWeb.findNext(goDown)
        }
    }

    /**
     * Executes search on string change
     */
    private inner class SearchExecutor : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
        override fun afterTextChanged(s: Editable) {
            binding.manContentWeb.findAllAsync(s.toString())
        }
    }

    private inner class FontChangeListener : OnSharedPreferenceChangeListener {
        override fun onSharedPreferenceChanged(prefs: SharedPreferences, key: String) {
            when (key) {
                Config.FONT_PREF_KEY -> binding.manContentWeb.settings.minimumFontSize = Config.fontSize
            }
        }
    }

    companion object {
        private const val PARAM_ADDRESS = "param.address"
        private const val PARAM_NAME = "param.name"

        fun newInstance(commandName: String, address: String): ManPageDialogFragment {
            val fragment = ManPageDialogFragment()
            val args = Bundle()
            args.putString(PARAM_ADDRESS, address)
            args.putString(PARAM_NAME, commandName)
            fragment.arguments = args
            return fragment
        }
    }
}